### PR TITLE
Fix sending post notification [MAILPOET-4685]

### PR DESCRIPTION
**Pull request by lysyjan**
## Description

This PR fixes issues with sending post notifications. There is a list of a few of them:
- Daily post notification sends out posts of the last many days instead of just the last day
- The same post notification is being sent over and over again
- The same post notification for a specific term is sent many times
- An old post notification was resent

## Code review notes

_N/A_

## QA notes

Please, focus on post notification and a variety of configurations. Try the behavior before the update and after it.

**Steps to reproduce:**
1. Set up a post notification with `1` post set to send `immediately`.
2. Create a post — the post notification should be received.
3. In `wp_mailpoet_newsletter_posts` delete the record with the post notification ID — this simulates the invalid data some users have.
4. Edit or create a page.
5. You’ll receive a post notification for that post from point 1 (= issue).

**Then (the issues):**
1. The notification history record may remain sending and in that case you won’t receive any further emails for that post notification, not even for legit new posts.
2. Or the post notification history record may become sent and now every new post or page edit will trigger the post notification again.

I don’t know why sometimes error 1. occurs (my local site or [here](https://secure.helpscout.net/conversation/2022809488/518554?folderId=59660) and [here](https://secure.helpscout.net/conversation/2023365837/518768?folderId=59658)) and sometimes error 2. (e.g. [here](https://secure.helpscout.net/conversation/2021549999/518142?folderId=1110911)).

This PR should fix both issues.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4685](https://mailpoet.atlassian.net/browse/MAILPOET-4685)

## After-merge notes

_N/A_

### CODE REVIEW

**Review by brezocordero — DISMISSED** [review]

File: mailpoet/lib/Config/Populator.php [hunk 1]
```diff
@@ -982,4 +983,32 @@ private function updateToUnifiedTrackingSettings() {
...
+      JOIN {$tasks} t ON q.task_id = t.id
+      SET n.status = :sentStatus
+      WHERE n.type = :type
+      AND n.status = :sendingStatus
+      AND t.status = :taskStatus
+    ", [
+      'type' => NewsletterEntity::TYPE_NOTIFICATION_HISTORY,
+      'sendingStatus' => NewsletterEntity::STATUS_SENDING,
+      'sentStatus' => NewsletterEntity::STATUS_SENT,
+      'taskStatus' => ScheduledTaskEntity::STATUS_COMPLETED,
```
Comment (brezocordero) [inline mailpoet/lib/Config/Populator.php:1009]: Just a note that the notification history will not have a value on `sent_at` but I don't see this as blocking.

File: mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php [hunk 2]
```diff
@@ -110,7 +110,7 @@ public function schedulePostNotification($postId) {
   }
 
   public function createPostNotificationSendingTask(NewsletterEntity $newsletter): ?ScheduledTaskEntity {
-    $notificationHistory = $this->newslettersRepository->findSendigNotificationHistoryWithPausedTask($newsletter);
+    $notificationHistory = $this->newslettersRepository->findSendingNotificationHistoryWithoutPausedTask($newsletter);
```
Comment (brezocordero) [inline mailpoet/lib/Newsletter/Scheduler/PostNotificationScheduler.php:113]: Thanks for changing the name.

File: mailpoet/lib/Cron/Workers/Scheduler.php [hunk 3]
```diff
@@ -211,6 +217,7 @@ public function processPostNotificationNewsletter($newsletter, SendingTask $queu
 
     // Because there is mixed usage of the old and new model, we want to be sure about the correct state
     $this->newslettersRepository->refresh($notificationHistory);
+    $this->sendingQueuesRepository->refresh($queue->getSendingQueueEntity());
```
Comment (brezocordero) [inline mailpoet/lib/Cron/Workers/Scheduler.php:220]: Great catch!

File: mailpoet/lib/Cron/Workers/SendingQueue/Tasks/Newsletter.php [hunk 4]
```diff
@@ -250,9 +250,9 @@ public function markNewsletterAsSent(NewsletterEntity $newsletter, Sending $send
       $newsletter->getType() === NewsletterEntity::TYPE_STANDARD ||
        $newsletter->getType() === NewsletterEntity::TYPE_NOTIFICATION_HISTORY
     ) {
-      $sendingQueue = $sendingTask->queue();
+      $scheduledTask = $sendingTask->task();
       $newsletter->setStatus(NewsletterEntity::STATUS_SENT);
-      $newsletter->setSentAt(new Carbon($sendingQueue->processedAt));
+      $newsletter->setSentAt(new Carbon($scheduledTask->processedAt));
```
Comment (brezocordero) [inline mailpoet/lib/Cron/Workers/SendingQueue/Tasks/Newsletter.php:255]: 🎩

**Review by brezocordero — APPROVED** [review]
The code looks good.
I have tested locally the following 2 cases:
- Site that had experienced the bug when upgrading to 3.99 and had a stuck notification, checking the notification history was updated as sent.
- Upgrading from 3.97 to this branch and checking that updating a page doesn’t resend existing post notifications.
- In both cases, that post notifications are sent when a new post is published.

**Comment by rodrigoprimo** [comment]
Merged to start preparing the emergency release as discussed in Slack.

**Pull request merged**